docs/rules/no-named-import.md

```diff
@@ -0,0 +1,87 @@
+# Prevent named imports from react (react/no-named-import)
+
+Prevent the use of named imports from React, so it's visible in code that it's React and not custom code. This rule
+also creates consistency inside a codebase because some people use named imports and some don't.
+
+## Rule Details
+
+### Options
+
+This rule has two different modes: 'import' and 'property' which can be set using the first argument. The second arguments accepts an array of strings, which are the allowed modules to be imported or accessed via property.
+
+```
+"react/no-named-import": [<enabled>, 'import' | 'property', []]
```

Comment:
what if someone wants to simultaneously require one thing be imported, and one thing be a property? For example, in my codebases I require hooks to be named imports, but I never permit `Component` to be anything except `React.Component`.

One possibility I'd expect here is a string argument, but the next option should be an object, whose keys are the names, and whose values are either "import" or "property".

Another would be an object with "import" and "property" properties, whose values are an array of names.

The advantage of the object key form is that it's trivial to ensure that each name only has one "mode" set; the advantage of the object array form is that it's much less boilerplate to write.

Perhaps there's a third alternative I'm not considering?

Reply:
@ljharb 
I see your point and thought about this too. We have three possibilities as I see it:

```
// current implementation
const options = ["error", "import", [/* <names> */]];
```

```
const options = ["error", {
  "import": ["Component"],
  "property": ["useEffect", "useState"],
}];
```

```
const options = ["error", {
  "useEffect": "import",
  "Component": "property",
}]
```

The current implementation automatically forces a mode for the non-specified names, which should not be the case after reading your comment. 

After giving it some thought I think the best way is to go for the 2nd implementation. It's also consistent with other rules that for example have a `forbid` option which is an array of strings.

What do you think?

Reply:
I'd expect, in both 2 and 3 in your comment, that the object would be a *third* argument - that the second argument would be a string so as to set the mode for all non-specified names, and the object would override some of those.

The reason I'm leaning towards the third is because it ensures that "foo" won't appear in both `import` and `property` - otherwise we have to do a runtime check and throw an error when there's overlap.

Reply:
@ljharb I've implemented the third option. Looking forward to your feedback